blockchain/chain/src/store/chain_store.rs

```diff
@@ -423,6 +434,80 @@ where
     })
 }
 
+pub fn get_chain_message<DB>(db: &DB, key: &Cid) -> Result<Box<dyn Message>, Error>
+where
+    DB: BlockStore,
+{
+    let value = db.read(key.key())?;
+    let bytes = value.ok_or_else(|| Error::UndefinedKey(key.to_string()))?;
+    let unsigned_message: Result<UnsignedMessage, _> = from_slice(&bytes);
+    if let Ok(s) = unsigned_message {
+        Ok(Box::new(s))
+    } else {
+        let signed_message: SignedMessage = from_slice(&bytes)?;
+        Ok(Box::new(signed_message))
+    }
+}
+
+pub fn message_for_tipset<DB>(db: &DB, ts: &Tipset) -> Result<Vec<Box<dyn Message>>, Error>
+where
+    DB: BlockStore,
+{
+    let mut applied: HashMap<Address, u64> = HashMap::new();
+    let mut balances: HashMap<Address, BigUint> = HashMap::new();
+    let state = StateTree::new_from_root(db, ts.parent_state())?;
+
+    //message to get all messages for block_header into a single iterator
+    type BoxMessage = Box<dyn Message>;
+    let mut get_message_for_block_header = |b: &BlockHeader| -> Result<Vec<BoxMessage>, Error> {
+        let (unsigned, signed) = block_messages(db, b)?;
+        let unsigned_box = unsigned.into_iter().map(|s| {
+            let box_msg: Box<dyn Message> = Box::new(s);
+            box_msg
+        });
+        let signed_box = signed.into_iter().map(|s| {
+            let box_msg: Box<dyn Message> = Box::new(s);
+            box_msg
+        });
+
+        let mut messages = Vec::new();
+        for message in unsigned_box.chain(signed_box) {
+            let from_address = message.from();
+            if let Some(_s) = applied.get(&from_address) {
+                let actor_state = state
+                    .get_actor(from_address)?
+                    .ok_or_else(|| Error::Other("Actor state not found".to_string()))?;
+                applied.insert(*from_address, actor_state.sequence);
+                balances.insert(*from_address, actor_state.balance);
+            }
+            let apply = applied.get(from_address);
+            if apply.map(|s| s != &message.sequence()).unwrap_or_default() {
+                continue;
+            }
+            let balance = balances.get(from_address).cloned();
+            if balance
+                .clone()
+                .map(|s| s < message.required_funds())
+                .unwrap_or_default()
+            {
+                continue;
+            }
+            balance.map(|s| balances.insert(*message.from(), s - message.required_funds()));
```

Comment:
```suggestion
            if let Some(bal) = balances.get_mut(from_address) {
                if *bal < message.required_funds() {
                    continue;
                }
                *bal -= message.required_funds();
            } else {
                continue;
            }
```
Avoids cloning the bigint and having to reinsert into hashmap (might also be able to be cleaner, but this is functionally what is needed)